WIP: Enable array api support in neighbor

yuejiaointel opened the pull request:
## Description

Follow up PR of https://github.com/uxlfoundation/scikit-learn-intelex/pull/2284 (will rebase after this one is merged)
 that refactor neighbors with array api standard

## Summary of Neighbor Code Refactoring Changes

### onedal layer changes:
- Removed validation methods from onedal estimators: `_validate_data`, `_get_weights`, `_validate_targets`, `_validate_n_classes` - all validation moved to sklearnex layer
- Removed unused imports: `Integral`, `_get_config`, and validation functions (`_check_array`, `_check_classification_targets`, `_check_X_y`, `_column_or_1d`, `_num_samples`)
- Classification target processing (class encoding, `outputs_2d_` setting) moved from onedal `_fit()` to sklearnex layer - onedal now expects pre-processed `_y` and `classes_` attributes
- Backend-specific formatting kept in onedal: GPU backend requires y in `(-1, 1)` shape for C++ compatibility
- `_kneighbors()` method updated with full array API support: uses `xp` namespace for all operations (`argsort`, `reshape`, `concatenate`, `arange`, `where`, `all`)
- Array API compatibility: changed `sample_mask[:, 0][dup_gr_nbrs] = False` to explicit operations using `xp.where()` and `xp.concatenate()` (array API doesn't support chained indexing assignment)
- `predict()` and `predict_proba()` methods removed from `KNeighborsClassifier` - computation logic moved to sklearnex
- `_predict_skl()` and `predict()` methods removed from `KNeighborsRegressor` - dispatch logic moved to sklearnex
- Table conversion order changed: convert to table FIRST, then get params from table (ensures dtype normalization from array API dtype to numpy dtype)
- `_predict_gpu()` kept in onedal for GPU backend support (called by sklearnex)

### sklearnex layer changes:
- Added `@enable_array_api` decorator to `KNeighborsClassifier` and `KNeighborsRegressor` (requires sklearn >= 1.5 for regressor due to `y_numeric` parameter)
- Created comprehensive validation and computation methods in `common.py`:
  - `_get_weights()` - adapted from sklearn with array API support (handles `dpctl`/`dpnp` arrays)
  - `_compute_weighted_prediction()` - regression prediction using array API `take()` and `stack()` (array API `take()` only supports 1-D indices)
  - `_compute_class_probabilities()` - classification probabilities with array API support (avoids fancy indexing via sample-by-sample accumulation)
  - `_predict_skl_regression()` and `_predict_skl_classification()` - unified prediction helpers that call `kneighbors()` and compute results
  - `_process_classification_targets()` - handles class encoding, `outputs_2d_` setting, validates n_classes >= 2
  - `_process_regression_targets()` - handles shape processing for regressors
  - `_kneighbors_post_processing()` - handles kd_tree sorting, query_is_train (X=None) case, return_distance decision
- `_onedal_fit()` in all estimators now: validates data, sets effective metric, processes targets, then calls onedal backend
- `_onedal_predict()` in classifier uses `_predict_skl_classification()` helper (handles X=None LOOCV case properly)
- `_onedal_predict_proba()` computes probabilities directly in sklearnex using `_compute_class_probabilities()`
- `_onedal_predict()` in regressor dispatches between GPU (`_predict_gpu()`) and SKL (`_predict_skl()`) paths based on device and weights
- `_onedal_kneighbors()` simplified in all estimators: validates X, calls onedal, returns result (post-processing removed - now in onedal)
- `_onedal_score()` in classifier uses `_transfer_to_host()` to convert array API arrays to numpy for sklearn's `accuracy_score()`
- Added `_validate_n_neighbors()`, `_set_effective_metric()`, `_validate_n_classes()`, `_kneighbors_validation()` helper methods
- `_fit_validation()` expanded with array API dtype support: `dtype=[xp.float64, xp.float32]`
- `_onedal_supported()` updated for array API: uses `get_namespace()` and `xp.asarray()` for type checks
- `kneighbors_graph()` uses `_transfer_to_host()` to ensure numpy arrays for scipy `csr_matrix` construction
- Metric aliases mapping added: `cityblock`→`manhattan`, `l1`→`manhattan`, `l2`→`euclidean` for oneDAL compatibility


### Testing changes:
- onedal tests (`onedal/neighbors/tests/test_knn_classification.py`) deprecated with notice pointing to sklearnex tests
- Tests moved to `sklearnex/neighbors/tests/test_neighbors.py`: `test_knn_classifier_iris()`, `test_knn_classifier_pickle()`
- Added `test_knn_classifier_single_class()` with `@pytest.mark.allow_sklearn_fallback` (oneDAL doesn't support one-class case)
- Removed queue parametrization from moved tests (sklearnex handles device selection internally)

### Documentation changes:
- Added `KNeighborsClassifier`, `KNeighborsRegressor`, `NearestNeighbors`, `LocalOutlierFactor` to `array_api.rst` showing array API support

### Architecture pattern:
- **onedal layer** = backend calls + full kneighbors post-processing + backend-specific formatting
- **sklearnex layer** = validation + target processing + dispatch + prediction computation
- Follows established pattern from PCA/SVM: validation in sklearnex, backend + post-processing in onedal

### Array API compliance:
- Cannot use chained indexing assignment: `array[:, 0][mask] = value` → explicit `xp.where()` + `xp.concatenate()`
- Cannot use fancy indexing for `__setitem__`: `proba_k[all_rows, idx] += weights` → sample-by-sample accumulation
- `take()` only supports 1-D indices: iterate over samples to gather neighbor values
- Use `_transfer_to_host()` before sklearn utility functions (accuracy_score, csr_matrix) that require numpy

---

<!--
PR should start as a draft, then move to ready for review state
after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, a PR with docs update doesn't require checkboxes for performance
while a PR with any change in actual code should list checkboxes and
justify how this code change is expected to affect performance (or justification should be self-evident).
-->

<details>
<summary>Checklist:</summary>

**Completeness and readability**

- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with updates and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least a summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance and/or quality metrics have changed or why changes are not expected.
- [x] I have extended the benchmarking suite and provided a corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

</details>

yuejiaointel commented on sklearnex/neighbors/_lof.py:
```diff
@@ -152,6 +152,18 @@ def _kneighbors(self, X=None, n_neighbors=None, return_distance=True):
+
+            from onedal.utils.validation import _check_array
+
+            X = _check_array(X, accept_sparse="csr", dtype=[np.float64, np.float32])
```
There are lots of parts in onedal neighbor we need to move to sklearnex, and I plan to do array api stuff after this step to avoid too much changes at once

david-cortes-intel replied:
I'd prefer to have all the changes in one PR as they'd be modifying the same parts.

david-cortes-intel commented on sklearnex/neighbors/_lof.py:
```diff
@@ -146,13 +144,20 @@ def _predict(self, X=None):
     @wrap_output_data
     def fit_predict(self, X, y=None):
-        return self.fit(X)._predict()
+        result = self.fit(X)._predict()
```
It looks like this and similar changes are not needed.

david-cortes-intel commented on sklearnex/neighbors/common.py:
```diff
@@ -24,21 +26,536 @@
+
+        return result_method
+
+    def _get_weights(self, dist, weights):
```
Does this mimic some method from scikit-learn? If so, could it call the original instead?

yuejiaointel commented:
/intelci: run

yuejiaointel commented:
/intelci: run

yuejiaointel commented:
/intelci: run

yuejiaointel commented:
/intelci: run

yuejiaointel commented:
/intelci: run

yuejiaointel commented:
/intelci: run

yuejiaointel commented:
/intelci: run

yuejiaointel commented:
/intelci: run

yuejiaointel commented:
/intelci: run

ethanglaser commented on onedal/neighbors/tests/test_knn_classification.py:
Can you remind me why it's necessary to remove this test entirely - for other array API updates we did not remove/modify onedal test files

yuejiaointel replied:
Good quesiton!
Based on [earlier comments](https://github.com/uxlfoundation/scikit-learn-intelex/pull/2700#discussion_r2726552428), since now the preprocessing is done at sklearnex level, the orig onedal test(test_knn_classification) is moved to sklearnex test(test_neighbors.py), and one of the test test_iris is just mathematics check and is removed since we already have conformance test

ethanglaser commented on onedal/neighbors/neighbors.py:
```diff
@@ -15,27 +15,17 @@
 from abc import ABCMeta, abstractmethod
-from numbers import Integral
 
 import numpy as np
```
This can be removed

ethanglaser commented on onedal/spmd/neighbors/neighbors.py:
No other onedal spmd files needed changes - why were these necessary? Can we remove some of these or adjust elsewhere?

ethanglaser reviewed (commented):
Mostly high level comments for now

yuejiaointel commented:
/intelci: run

yuejiaointel commented on onedal/neighbors/neighbors.py:
```diff
@@ -176,91 +106,63 @@ def __init__(
-        self.classes_ = None
+        # REFACTOR: Shape processing moved to sklearnex layer
+        # _shape should be set by _process_classification_targets or _process_regression_targets in sklearnex
+        # self._shape = None
```
thx, removed!

yuejiaointel commented:
/intelci: run

yuejiaointel commented:
/intelci: run

david-cortes-intel commented on doc/sources/array_api.rst:
```diff
@@ -105,6 +105,10 @@ The following patched classes have support for array API inputs:
+- :obj:`sklearn.neighbors.KNeighborsClassifier`
+- :obj:`sklearn.neighbors.KNeighborsRegressor`
+- :obj:`sklearn.neighbors.NearestNeighbors`
+- :obj:`sklearn.neighbors.LocalOutlierFactor`
```
@yuejiaointel @ethanglaser I have it a try and both onedal and sklearnex report that it runs on GPU:
```
INFO:sklearnex: sklearn.neighbors.LocalOutlierFactor.fit: running accelerated version on GPU
```
```
-----------------------------------------------------------------------------
File: cpp/oneapi/dal/backend/primitives/search/search_dpc.cpp, Line: 48
Profiler task_name: search.treat_indices, q
search.treat_indices 24.89ms
search.squared_l2.inner_blocking 249.78ms
-----------------------------------------------------------------------------
File: cpp/oneapi/dal/backend/primitives/search/search_dpc.cpp, Line: 608
Profiler task_name: selection_loop.selection, this->get_queue()
-----------------------------------------------------------------------------
```

But some things I'm not clear whether they are CPU or GPU:
```
-----------------------------------------------------------------------------
File: [./cpp/oneapi/dal/backend/primitives/reduction/reduction.hpp](https://vscode-remote+ssh-002dremote-002bsdpdal630226-002ejf-002eintel-002ecom.vscode-resource.vscode-cdn.net/export/users/dcortes/repos/sklex-dpnp/cpp/oneapi/dal/backend/primitives/reduction/reduction.hpp), Line: 60
Profiler task_name: reduction.reduce_by_rows, q
reduction.reduce_by_rows 40.87ms
-----------------------------------------------------------------------------
File: [./cpp/oneapi/dal/backend/primitives/reduction/reduction.hpp](https://vscode-remote+ssh-002dremote-002bsdpdal630226-002ejf-002eintel-002ecom.vscode-resource.vscode-cdn.net/export/users/dcortes/repos/sklex-dpnp/cpp/oneapi/dal/backend/primitives/reduction/reduction.hpp), Line: 60
Profiler task_name: reduction.reduce_by_rows, q
reduction.reduce_by_rows 11.97us
-----------------------------------------------------------------------------
```

@Alexandr-Solovev @Vika-F Could you comment on what's the status of GPU support for LocalOutlierFactor?

david-cortes-intel commented on sklearnex/neighbors/knn_classification.py:
```diff
@@ -149,23 +182,127 @@ def _onedal_fit(self, X, y, queue=None):
+            _check_classification_targets(y)
+
+        # Process classes - note: np.unique is used for class extraction
+        # This is acceptable as classes are typically numpy arrays in sklearn
```
This is not the case. They can also be other things like pd.Series, pl.Series, etc.

david-cortes-intel commented:
@yuejiaointel I'm getting an error when trying to call `kneighbors` on GPU:
```python
import os
os.environ["SCIPY_ARRAY_API"] = "1"

import numpy as np
import dpnp

rng = np.random.default_rng(seed=123)
X = rng.standard_normal(size=(1000, 20), dtype=np.float32)
Xd = dpnp.array(X, device="gpu")

from sklearnex import config_context
from sklearnex.neighbors import NearestNeighbors

with config_context(array_api_dispatch=True):
    nn = NearestNeighbors().fit(Xd)
    neigh = nn.kneighbors(Xd)
```
```
File /localdisk2/mkl/dcortes/repos/sklex-dpnp/sklearnex/_device_offload.py:191, in wrap_output_data.<locals>.wrapper(self, *args, **kwargs)
    186 if (
    187     usm_iface := getattr(data, "__sycl_usm_array_interface__", None)
    188 ) and not hasattr(result, "__sycl_usm_array_interface__"):
    189     queue = usm_iface["syclobj"]
    190     return (
--> [191](https://vscode-remote+ssh-002dremote-002bsdpdal630226-002ejf-002eintel-002ecom.vscode-resource.vscode-cdn.net/localdisk2/mkl/dcortes/repos/sklex-dpnp/sklearnex/_device_offload.py:191)         copy_to_dpnp(queue, result)
    192         if is_dpnp_ndarray(data)
    193         else copy_to_usm(queue, result)
    194     )
...
    421 if self.ndim == 0:
--> [422](https://vscode-remote+ssh-002dremote-002bsdpdal630226-002ejf-002eintel-002ecom.vscode-resource.vscode-cdn.net/localdisk2/mkl/dcortes/miniforge3/envs/dpnp18/lib/python3.13/site-packages/dpnp/dpnp_array.py:422)     raise TypeError("iteration over a 0-d array")
    423 return (self[i] for i in range(self.shape[0]))

TypeError: iteration over a 0-d array
```

yuejiaointel commented:
/intelci: run ml-benchmarks
set get-build=f106e0bf-84d3-f1cf-ac46-a4bf010d0e2d

homksei commented:
/intelci: run ml-benchmarks
set get-build=f106e0bf-84d3-f1cf-ac46-a4bf010d0e2d

david-cortes-intel commented:
@yuejiaointel After the fixes, calling `kneighbors_graph` returns something with float64 type regardless of the input dtype:
```python
import os, sys
os.environ["SCIPY_ARRAY_API"] = "1"

import numpy as np
import dpnp

rng = np.random.default_rng(seed=123)
X = rng.standard_normal(size=(1000, 20), dtype=np.float32)
Xd = dpnp.array(X, device="gpu")

from sklearnex import config_context
from sklearnex.neighbors import NearestNeighbors

with config_context(array_api_dispatch=True):
    nn = NearestNeighbors().fit(Xd)
    neigh = nn.kneighbors_graph(Xd)
```

yuejiaointel commented on sklearnex/neighbors/knn_regression.py:
```diff
@@ -130,20 +156,104 @@ def _onedal_fit(self, X, y, queue=None):
+        """GPU prediction path - calls onedal backend."""
+        if X is not None:
+            xp, _ = get_namespace(X)
+            if getattr(self, "effective_metric_", self.metric) == "precomputed":
```
right removed, onedal supported already reject precomputed, so predict gpu can never be called with precompted

ethanglaser commented:
onedal/_device_offload.py should be restored before merge. #2940 will be merged after all array API updates are complete

ethanglaser commented on sklearnex/neighbors/knn_classification.py:
```diff
@@ -125,9 +130,15 @@ def score(self, X, y, sample_weight=None):
+
+        # Validate kneighbors parameters (inherited from KNeighborsDispatchingBase)
+        self._kneighbors_validation(X, n_neighbors)
+
```
These should probably go inside _onedal_kneighbors assuming sklearn has their own set of validation for n_neighbors. Also since it looks like _onedal_kneighbors is the exact same across the different knns, maybe it should be moved to common.py

yuejiaointel commented:
/intelci: run ml-benchmarks
set get-build=f107da71-d86c-f1d9-9a7e-a4bf010d0e2d

ethanglaser commented:
> @yuejiaointel Getting another error with torch:
> 
> ```python
> import os
> os.environ["SCIPY_ARRAY_API"] = "1"
> 
> import numpy as np
> import torch
> 
> rng = np.random.default_rng(seed=123)
> X = rng.standard_normal(size=(1000, 20), dtype=np.float32)
> Xt = torch.tensor(X, device="xpu")
> 
> from sklearnex import config_context
> from sklearnex.neighbors import NearestNeighbors
> 
> with config_context(array_api_dispatch=True):
>     nn = NearestNeighbors().fit(Xt)
>     neigh = nn.kneighbors_graph(Xt)
> ```
> 
> ```
> TypeError: Object of type 'torch.dtype' is not an instance of 'dtype'
> ```

Just curious why torch here? We don't have any torch validation of our array API estimators as far as I know - would others not fail as well?

yuejiaointel commented:
> ```python
> import os
> os.environ["SCIPY_ARRAY_API"] = "1"
> 
> import numpy as np
> import torch
> 
> rng = np.random.default_rng(seed=123)
> X = rng.standard_normal(size=(1000, 20), dtype=np.float32)
> Xt = torch.tensor(X, device="xpu")
> y = rng.integers(3, size=X.shape[0])
> yt = torch.tensor(y, device="xpu")
> 
> from sklearnex import config_context
> from sklearnex.neighbors import KNeighborsClassifier
> 
> with config_context(array_api_dispatch=True):
>     nn = KNeighborsClassifier().fit(Xt, yt)
>     probs = nn.predict_proba(Xt)
> ```

fixed! I added a __sycl_usm_array_interface__ check on onedal_supported to reject any non sycl arrays and let them fall back to sklearn
output I see now:
build) yuejiao@sdpdal630226:/export/users/yuejiao$  cd /export/users/yuejiao && python -c "
> import os
> os.environ['SCIPY_ARRAY_API'] = '1'
> import numpy as np
> import torch
> rng = np.random.default_rng(seed=123)
> X = rng.standard_normal(size=(1000, 20), dtype=np.float32)
> Xt = torch.tensor(X, device='xpu')
> y = rng.integers(3, size=X.shape[0])
> yt = torch.tensor(y, device='xpu')
> from sklearnex import config_context
> from sklearnex.neighbors import KNeighborsClassifier
> with config_context(array_api_dispatch=True):
>     nn = KNeighborsClassifier().fit(Xt, yt)
>     probs = nn.predict_proba(Xt)
>     print(f'Success! probs shape: {probs.shape}, dtype: {probs.dtype}')
> "
/export/users/yuejiao/builds/knn_01_12_1728/__release_lnx/daal4py-3.12/sklearnex/_device_offload.py:160: UserWarning: Some metric_kwargs have been passed ({'p': 2}) but aren't usable for this case (EuclideanArgKmin64) and will be ignored.
  return branches["sklearn"](obj, *hostargs, **hostkwargs)
Success! probs shape: (1000, 3), dtype: float64

yuejiaointel commented:
> @ethanglaser @yuejiaointel I see now KNN estimators are being tested with the same methods and inputs as both 'standard estimator' and 'special estimator' here: https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/sklearnex/tests/test_patching.py https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/sklearnex/tests/test_run_to_run_stability.py
> 
> It looks like the only difference is that the special ones call it with `algorithm="brute"`, while the standard ones do not specify the algorithm:
> 
> https://github.com/uxlfoundation/scikit-learn-intelex/blob/bf43701884315a79a76810cbdc24f9bddcb4b9fe/sklearnex/tests/utils/base.py#L134
> 
> Is that because GPU support requires `algorithm="brute"`, or is there some other reason why it needs to be under the "special" list?
> 
> ~And if this is the case, should KMeans also get added to the special list with the algorithm that is supported on GPU?~ Actually I now see that it has this note for GPU:
> 
> > ‘elkan’ falls back to ‘lloyd’

I think we need both standard and special because the standard (algorithem = "auto") would pick kd_tree with the test data (euclidean metric)
```python
 if self._fit_method in ["auto", "ball_tree"]:
            condition = (
                self.n_neighbors is not None
                and self.n_neighbors >= self.n_samples_fit_ // 2
            )
            if self.n_features_in_ > 15 or condition:
                result_method = "brute"
            else:
                if self.effective_metric_ in ["euclidean"]:
                    result_method = "kd_tree"
                else:
                    result_method = "brute"
        else:
            result_method = self._fit_method
 ```
 and algorithm = "brute" will not be tested so we need to test it in special instance

yuejiaointel commented:
/intelci: run ml-benchmarks
set get-build=f1086a06-aecb-f15c-a7c9-a4bf010d0e2d

yuejiaointel commented:
/intelci: run ml-benchmarks
set get-build=f1086a06-aecb-f15c-a7c9-a4bf010d0e2d

yuejiaointel commented:
/intelci: run ml-benchmarks
set get-build=f1086a06-aecb-f15c-a7c9-a4bf010d0e2d

david-cortes-intel commented on sklearnex/_device_offload.py:
```diff
@@ -183,9 +183,15 @@ def wrapper(self, *args, **kwargs) -> Any:
             # Remove check for result __sycl_usm_array_interface__ on deprecation of use_raw_inputs
+            # For tuple/list results (e.g. kneighbors), check elements instead of the container
+            result_on_device = (
+                all(hasattr(r, "__sycl_usm_array_interface__") for r in result)
```
Same comment here - please avoid checking or using this attribute.

yuejiaointel commented:
> > > @yuejiaointel Getting another error with torch:
> > > ```python
> > > import os
> > > os.environ["SCIPY_ARRAY_API"] = "1"
> > > 
> > > import numpy as np
> > > import torch
> > > 
> > > rng = np.random.default_rng(seed=123)
> > > X = rng.standard_normal(size=(1000, 20), dtype=np.float32)
> > > Xt = torch.tensor(X, device="xpu")
> > > 
> > > from sklearnex import config_context
> > > from sklearnex.neighbors import NearestNeighbors
> > > 
> > > with config_context(array_api_dispatch=True):
> > >     nn = NearestNeighbors().fit(Xt)
> > >     neigh = nn.kneighbors_graph(Xt)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > ```
> > > TypeError: Object of type 'torch.dtype' is not an instance of 'dtype'
> > > ```
> > 
> > 
> > Fixed with same fix above python -c "
> > > import os
> > > os.environ['SCIPY_ARRAY_API'] = '1'
> > > import numpy as np
> > > import torch
> > > rng = np.random.default_rng(seed=123)
> > > X = rng.standard_normal(size=(1000, 20), dtype=np.float32)
> > > Xt = torch.tensor(X, device='xpu')
> > > from sklearnex import config_context
> > > from sklearnex.neighbors import NearestNeighbors
> > > with config_context(array_api_dispatch=True):
> > > nn = NearestNeighbors().fit(Xt)
> > > neigh = nn.kneighbors_graph(Xt)
> > > print(f'Success! type: {type(neigh)}, shape: {neigh.shape}, dtype: {neigh.dtype}')
> > > "
> > > Success! type: <class 'scipy.sparse._csr.csr_matrix'>, shape: (1000, 1000), dtype: float32
> 
> This is now falling back to sklearn. It should execute on GPU.

fixed! now I can see it says when using  SKLEARNEX_VERBOSE=INFO 
INFO:sklearnex: sklearn.neighbors.NearestNeighbors.fit: running accelerated version on GPU
INFO:sklearnex: sklearn.neighbors.NearestNeighbors.kneighbors: running accelerated version on GPU

yuejiaointel commented:
/intelci: run

david-cortes-intel commented:
> you can see here after check_array we have float64 even tho X is float32, for numpy we preserved the datatype but for dataframe it is just choosing the first dtype in list. One potential fix would be convert plars df to numpy before passing to check_array, but it seems to be beyond the scope of this pr

Indeed, looks like something is not working correctly across several estimators. Scikit-learn would preserve the dtype in those cases:
```python
import numpy as np
rng = np.random.default_rng(seed=123)
X = rng.standard_normal(size=(1000, 20), dtype=np.float32)
y = rng.standard_normal(size=X.shape[0]).astype(np.float32)
from sklearn import config_context
from sklearn.linear_model import LinearRegression
import polars as pl
Xdf = pl.DataFrame(X)
ys = pl.Series(y)
lr = LinearRegression().fit(Xdf, ys)
print('coef_ dtype:', lr.coef_.dtype)
pred = lr.predict(Xdf)
print('predict dtype:', pred.dtype)
```
```
coef_ dtype: float32
predict dtype: float32
```

yuejiaointel commented on sklearnex/_device_offload.py:
```diff
@@ -182,11 +182,10 @@ def wrapper(self, *args, **kwargs) -> Any:
-                usm_iface := getattr(data, "__sycl_usm_array_interface__", None)
-            ) and not hasattr(result, "__sycl_usm_array_interface__"):
-                queue = usm_iface["syclobj"]
+            sycl_type, _, _ = _get_sycl_namespace(data)
```
It was from earlier comment, but public test seems to be working for orig code, running the entire test to confirm

yuejiaointel commented:
/intelci: run

david-cortes-intel commented:
> > @yuejiaointel I'm getting an error on LocalOutlierFactor with torch:
> > ```
> > import os
> > os.environ["SCIPY_ARRAY_API"] = "1"
> > 
> > import numpy as np
> > import torch
> > 
> > rng = np.random.default_rng(seed=123)
> > X = rng.standard_normal(size=(1000, 20), dtype=np.float32)
> > Xd = torch.tensor(X, device="xpu")
> > 
> > from sklearnex import config_context
> > from sklearnex.neighbors import LocalOutlierFactor
> > 
> > with config_context(array_api_dispatch=True):
> >     model = LocalOutlierFactor(novelty=True).fit(Xd)
> >     pred = model.predict(Xd)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > File /localdisk2/mkl/dcortes/repos/sklex-dpnp/sklearnex/neighbors/_lof.py:146, in LocalOutlierFactor._predict(self, X)
> >     144     xp, _ = get_namespace(X)
> >     145     output = self.decision_function(X) < 0
> > --> [146](https://vscode-remote+ssh-002dremote-002bsdpdal630226-002ejf-002eintel-002ecom.vscode-resource.vscode-cdn.net/localdisk2/mkl/dcortes/repos/sklex-dpnp/sklearnex/neighbors/_lof.py:146)     is_inlier = xp.ones_like(output, dtype=int)
> >     147     is_inlier[output] = -1
> >     148 else:
> > 
> > TypeError: ones_like(): argument 'input' (position 1) must be Tensor, not numpy.ndarray
> > ```
> 
> Thx, fixed!

@yuejiaointel It works now, but outputs numpy. It should output torch.

david-cortes-intel commented:
@yuejiaointel Also please look into why the failing case above is not catched by the current tests. From what I see, it should execute some kind of test like that:
https://github.com/uxlfoundation/scikit-learn-intelex/blob/f43082d2e0658ac7fc1002084cec16b129caf845/sklearnex/tests/utils/base.py#L131

yuejiaointel commented:
> > > @yuejiaointel I'm getting an error on LocalOutlierFactor with torch:
> > > ```
> > > import os
> > > os.environ["SCIPY_ARRAY_API"] = "1"
> > > 
> > > import numpy as np
> > > import torch
> > > 
> > > rng = np.random.default_rng(seed=123)
> > > X = rng.standard_normal(size=(1000, 20), dtype=np.float32)
> > > Xd = torch.tensor(X, device="xpu")
> > > 
> > > from sklearnex import config_context
> > > from sklearnex.neighbors import LocalOutlierFactor
> > > 
> > > with config_context(array_api_dispatch=True):
> > >     model = LocalOutlierFactor(novelty=True).fit(Xd)
> > >     pred = model.predict(Xd)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > ```
> > > File /localdisk2/mkl/dcortes/repos/sklex-dpnp/sklearnex/neighbors/_lof.py:146, in LocalOutlierFactor._predict(self, X)
> > >     144     xp, _ = get_namespace(X)
> > >     145     output = self.decision_function(X) < 0
> > > --> [146](https://vscode-remote+ssh-002dremote-002bsdpdal630226-002ejf-002eintel-002ecom.vscode-resource.vscode-cdn.net/localdisk2/mkl/dcortes/repos/sklex-dpnp/sklearnex/neighbors/_lof.py:146)     is_inlier = xp.ones_like(output, dtype=int)
> > >     147     is_inlier[output] = -1
> > >     148 else:
> > > 
> > > TypeError: ones_like(): argument 'input' (position 1) must be Tensor, not numpy.ndarray
> > > ```
> > 
> > 
> > Thx, fixed!
> 
> @yuejiaointel It works now, but outputs numpy. It should output torch.

fixed! it was because dispatch logic was converting torch to numpy internally, added a xp.asarray to convert it back at the end

david-cortes-intel commented on sklearnex/neighbors/knn_classification.py:
```diff
@@ -105,12 +112,15 @@ def predict_proba(self, X):
             },
             X,
         )
+        # Convert result back to the input array namespace (e.g. torch).
```
It's not supposed to work like that. Please modify the underlying function so that it would not transfer data to host unless strictly needed, which I gather from the comments should only be in SPMD mode:
https://github.com/yuejiaointel/scikit-learn-intelex/blob/68f7b5004391e7b0cd00005443a939abb0f99322/sklearnex/neighbors/common.py#L173

Also in the case of SPMD, if it's transferring data to host, wouldn't that defy the purpose of SPMD which is to fit to datasets larger than a single node can host? CC @Alexandr-Solovev

david-cortes-intel commented on sklearnex/neighbors/_lof.py:
```diff
@@ -122,20 +136,17 @@ def fit(self, X, y=None):
-            is_inlier = xp.ones_like(output, dtype=int)
-            is_inlier[output] = -1
+            xp, _ = get_namespace(output)
+            is_inlier = xp.where(output, xp.asarray(-1), xp.asarray(1))
```
Looks like this produces int64. Please use a shorter type, since these are binary. Int32 should be fine.

david-cortes-intel commented:
> > @yuejiaointel Also please look into why the failing case above is not catched by the current tests. From what I see, it should execute some kind of test like that:
> > https://github.com/uxlfoundation/scikit-learn-intelex/blob/f43082d2e0658ac7fc1002084cec16b129caf845/sklearnex/tests/utils/base.py#L131
> > 
> > https://github.com/uxlfoundation/scikit-learn-intelex/blob/f43082d2e0658ac7fc1002084cec16b129caf845/sklearnex/tests/test_patching.py#L231
> 
> After some investigation I think it is related to @available_if

But that should return `True` for `predict` when passing `novelty=True` as done in the special estimator list.

yuejiaointel commented on sklearnex/neighbors/_lof.py:
```diff
@@ -122,20 +136,17 @@ def fit(self, X, y=None):
+            xp, _ = get_namespace(output)
+            is_inlier = xp.where(output, xp.asarray(-1), xp.asarray(1))
         else:
             is_inlier = np.ones(self.n_samples_fit_, dtype=int)
```
fixed! changed to xp.int32

yuejiaointel commented:
> @yuejiaointel My bad, it appears this is not how it's supposed to be.
> 
> I see by now scikit-learn has one classifier with array API support which we can use as a reference.
> 
> If passing array API inputs that are not numpy, then the result from predict is returned with the same dtype as the input (e.g. float32):
> 
> ```python
> import os
> os.environ["SCIPY_ARRAY_API"] = "1"
> import numpy as np
> import array_api_strict
> rng = np.random.default_rng(seed=123)
> X = rng.standard_normal(size=(1000, 20))
> Xs = array_api_strict.asarray(X, dtype=array_api_strict.float32)
> y = rng.integers(2, size=X.shape[0])
> ys = array_api_strict.asarray(y, dtype=array_api_strict.float32)
> 
> from sklearn import config_context
> from sklearn.linear_model import RidgeClassifier
> with config_context(array_api_dispatch=True):
>     pred = RidgeClassifier(solver="svd").fit(Xs, ys).predict(Xs)
> pred.dtype
> ```
> 
> ```
> array_api_strict.float32
> ```
> 
> ```python
> X64 = array_api_strict.asarray(X, dtype=array_api_strict.float64)
> y64 = array_api_strict.asarray(y, dtype=array_api_strict.float64)
> 
> with config_context(array_api_dispatch=True):
>     pred = RidgeClassifier(solver="svd").fit(X64, y64).predict(X64)
> pred.dtype
> ```
> 
> ```
> array_api_strict.float64
> ```
> 
> The same applies if passing NumPy inputs that are not float64, whether under array API dispatching or not:
> 
> ```python
> with config_context(array_api_dispatch=True):
>     pred = RidgeClassifier(solver="svd").fit(X.astype(np.float32), y.astype(np.float32)).predict(X.astype(np.float32))
> pred.dtype
> ```
> 
> ```
> dtype('float32')
> ```
> 
> ```python
> pred = RidgeClassifier(solver="svd").fit(X.astype(np.float32), y.astype(np.float32)).predict(X.astype(np.float32))
> pred.dtype
> ```
> 
> ```
> dtype('float32')
> ```
> 
> ```python
> pred = RidgeClassifier(solver="svd").fit(X.astype(np.int32), y.astype(np.int32)).predict(X.astype(np.int32))
> pred.dtype
> ```
> 
> ```
> dtype('int32')
> ```
> 
> But if the input is NumPy float64 or Pandas float64, then the result is returned as int64:
> 
> ```python
> with config_context(array_api_dispatch=True):
>     pred = RidgeClassifier(solver="svd").fit(X, y).predict(X)
> pred.dtype
> ```
> 
> ```
> dtype('int64')
> ```
> 
> ```python
> pred = RidgeClassifier(solver="svd").fit(pd.DataFrame(X), pd.Series(y)).predict(pd.DataFrame(X))
> pred.dtype
> ```
> 
> ```
> dtype('int64')
> ```
> 
> It appears their logic is very inconsistent, but since by their own documentation, prediction methods shouls follow a design of 'everything follows X': https://scikit-learn.org/stable/modules/array_api.html
> 
> Then I'd say the most sensible handling here would be:
> 
>     * Return xp dtype if using array API.
> 
>     * Return int64 otherwise.

thx for this explaination! fixed to use xp.dtype for array api otherwise int64

yuejiaointel commented on sklearnex/neighbors/common.py:
```diff
@@ -26,18 +27,205 @@
+
+        probabilities = []
+        for k, classes_k in enumerate(classes_):
+            pred_labels = _y[:, k][neigh_ind]
```
changing to xp.take will reduce performance by 20%

david-cortes-intel replied:
Are you able to make up a minimal example where `xp.take` is slower than bracketed indexing?

david-cortes-intel commented on onedal/neighbors/neighbors.py:
```diff
@@ -15,27 +15,16 @@
 from onedal.common._backend import bind_default_backend
 from onedal.utils import _sycl_queue_manager as QM
+from sklearnex.utils._array_api import get_namespace
 
```
I think the bot is right here. Other array API classes in the onedal module are not importing `get_namespace` from sklearnex.

david-cortes-intel commented on sklearnex/neighbors/common.py:
```diff
@@ -26,18 +27,213 @@
+                for c in range(n_classes):
+                    mask = pred_labels == c
+                    mask = xp.astype(mask, neigh_dist.dtype)
+                    proba_k[:, c] = xp.sum(mask * weights_k, axis=1)
```
Is this faster than slicing weights_k with `mask`?

david-cortes-intel commented on sklearnex/neighbors/knn_classification.py:
```diff
@@ -149,23 +181,135 @@ def _onedal_fit(self, X, y, queue=None):
+        xp, _ = get_namespace(X)
+        responses = from_table(result.responses, like=X)
+        return xp.take(
+            self.classes_, xp.asarray(xp.reshape(responses, (-1,)), dtype=xp.int64)
```
Also to clarify: there's a difference between LOF which has 'X' but no 'y' and KNNClassifier, which explicitly takes 'y'. When there is a 'y' involved in a classifier, `.predict()` should return values like the original 'y'. The original 'y' is not required to be encoded integers, can be 'str' and similar.

david-cortes-intel commented on sklearnex/neighbors/_lof.py:
```diff
@@ -122,20 +136,21 @@ def fit(self, X, y=None):
+            is_inlier = xp.where(output, -ones, ones)
         else:
-            is_inlier = np.ones(self.n_samples_fit_, dtype=int)
+            is_inlier = np.ones(self.n_samples_fit_, dtype=np.int64)
```
Yes, but I mean: it shouldn't return numpy for array API.

yuejiaointel replied:
Fixed! now I see this

 python -c "
import os
os.environ['SCIPY_ARRAY_API'] = '1'

import numpy as np
import torch

rng = np.random.default_rng(seed=123)
X = rng.standard_normal(size=(1000, 20), dtype=np.float32)
Xt = torch.tensor(X, device='xpu')

from sklearnex import config_context
from sklearnex.neighbors import LocalOutlierFactor

with config_context(array_api_dispatch=True):
    pred = LocalOutlierFactor().fit_predict(Xt)

print(f'type: {type(pred)}')
print(f'device: {pred.device}')
print(f'dtype: {pred.dtype}')
print(f'unique: {torch.unique(pred)}')
"

type: <class 'torch.Tensor'>
device: xpu:0
dtype: torch.int64
unique: tensor([-1,  1], device='xpu:0')

yuejiaointel commented:
/intelci: run

yuejiaointel commented:
> @yuejiaointel Calling `.predict_proba()` in the SPMD version of KNN classifier throws an error:
> 
> ```
> RuntimeError: Executing functions from SPMD backend requires a queue
> ```
> 
> Error can be triggered by adding a line like this in the [SPMD example](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/examples/sklearnex/knn_bf_classification_spmd.py):
> 
> ```python
> prob = model_spmd.predict_proba(dpnp_X_test)
> ```
> 
> Since SPMD estimators are not meant to fall back to sklearn given that it doesn't support distributed mode, maybe the most reasonable option here would be to error out when trying to call something that's not supported in SPMD.
> 
> But `.predict_proba()` is a bit more tricky since it's a base method. Maybe the old approach of transferring all the data to one rank could be kept, and we can leave an open issue about changing that to mpi4py's `Isendrecv_replace`.



> @yuejiaointel Calling `.predict_proba()` in the SPMD version of KNN classifier throws an error:
> 
> ```
> RuntimeError: Executing functions from SPMD backend requires a queue
> ```
> 
> Error can be triggered by adding a line like this in the [SPMD example](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/examples/sklearnex/knn_bf_classification_spmd.py):
> 
> ```python
> prob = model_spmd.predict_proba(dpnp_X_test)
> ```
> 
> Since SPMD estimators are not meant to fall back to sklearn given that it doesn't support distributed mode, maybe the most reasonable option here would be to error out when trying to call something that's not supported in SPMD.
> 
> But `.predict_proba()` is a bit more tricky since it's a base method. Maybe the old approach of transferring all the data to one rank could be kept, and we can leave an open issue about changing that to mpi4py's `Isendrecv_replace`.


Thx, that is a good test, added the raise error in spmd class and it is works

yuejiaointel commented:
Here is what has been tested:
```
Neighbor Estimators — Complete Method Matrix

1. NearestNeighbors
+--------------------------+--------+-----+---------------+----------+------+-----------------+-------------------+
| Method                   | onedal | GPU | CPU           | sklearn  | SPMD | _convert_result | @wrap_output_data |
+--------------------------+--------+-----+---------------+----------+------+-----------------+-------------------+
| fit                      | Yes    | Yes | Yes           | Yes      | Yes  | No              | No                |
| kneighbors               | Yes    | Yes | Yes           | Yes      | Yes  | Yes             | No                |
| kneighbors_graph         | via kn | Yes | Yes           | Yes      | Yes  | No              | No                |
| radius_neighbors         | No     | -   | -             | Yes      | No   | No              | Yes               |
| radius_neighbors_graph   | No     | -   | -             | Yes      | No   | No              | No                |
+--------------------------+--------+-----+---------------+----------+------+-----------------+-------------------+

2. KNeighborsClassifier
+--------------------------+--------+-----+---------------+----------+------+-----------------+-------------------+
| Method                   | onedal | GPU | CPU           | sklearn  | SPMD | _convert_result | @wrap_output_data |
+--------------------------+--------+-----+---------------+----------+------+-----------------+-------------------+
| fit                      | Yes    | Yes | Yes           | Yes      | Yes  | No              | No                |
| predict                  | Yes    | Yes | Yes           | Yes      | Yes  | Yes             | Yes               |
| predict_proba            | Yes    | Yes | Yes           | Yes      | NIE  | Yes             | Yes               |
| score                    | Yes    | Yes | Yes           | Yes      | Yes  | No              | Yes               |
| kneighbors               | Yes    | Yes | Yes           | Yes      | Yes  | Yes             | No                |
| kneighbors_graph         | via kn | Yes | Yes           | Yes      | Yes  | No              | No                |
+--------------------------+--------+-----+---------------+----------+------+-----------------+-------------------+
NIE = raises NotImplementedError

3. KNeighborsRegressor
+--------------------------+--------+-----+---------------+----------+------+-----------------+-------------------+
| Method                   | onedal | GPU | CPU           | sklearn  | SPMD | _convert_result | @wrap_output_data |
+--------------------------+--------+-----+---------------+----------+------+-----------------+-------------------+
| fit                      | Yes    | Yes | Yes           | Yes      | Yes  | No              | No                |
| predict                  | Yes    | Yes | Yes           | Yes      | Yes  | Yes             | Yes               |
| score                    | Yes    | Yes | Yes           | Yes      | Yes  | No              | Yes               |
| kneighbors               | Yes    | Yes | Yes           | Yes      | Yes  | Yes             | No                |
| kneighbors_graph         | via kn | Yes | Yes           | Yes      | Yes  | No              | No                |
+--------------------------+--------+-----+---------------+----------+------+-----------------+-------------------+

4. LocalOutlierFactor
+--------------------------+--------+-----+---------------+----------+------+-----------------+-------------------+
| Method                   | onedal | GPU | CPU           | sklearn  | SPMD | _convert_result | @wrap_output_data |
+--------------------------+--------+-----+---------------+----------+------+-----------------+-------------------+
| fit                      | Yes    | Yes | Yes           | Yes      | N/A  | No              | No                |
| fit_predict              | via ft | Yes | Yes           | Yes      | N/A  | Yes             | Yes               |
| kneighbors               | Yes    | Yes | Yes           | Yes      | N/A  | Yes             | No                |
| score_samples (novelty)  | Yes    | Yes | Yes           | Yes      | N/A  | Yes             | Yes               |
| predict (novelty)        | inher  | ind | ind           | Yes      | N/A  | No              | No                |
| decision_function (nov)  | inher  | ind | ind           | Yes      | N/A  | No              | No                |
+--------------------------+--------+-----+---------------+----------+------+-----------------+-------------------+
N/A = no SPMD subclass for LOF
ind = indirectly via kneighbors chain
inher = inherited from sklearn

GPU vs CPU Restrictions
+------------------------------------------+-----+-----+
| Condition                                | GPU | CPU |
+------------------------------------------+-----+-----+
| algorithm="brute"                        | Yes | Yes |
| algorithm="kd_tree"                      | No  | Yes |
| algorithm="ball_tree"                    | No  | No  |
| brute metrics: minkowski/euclidean/      | Yes | Yes |
|   manhattan/chebyshev/cosine             |     |     |
| kd_tree metrics: euclidean only          | -   | Yes |
| sparse input                            | No  | No  |
| multi-output                            | No  | No  |
+------------------------------------------+-----+-----+

Input Types Tested (91/91 passed, all 4 estimators x all methods)
+--------+--------+-----+---------+---------------+------------+---------------+-------------+
| Type   | Device | fit | predict | predict_proba | kneighbors | score_samples | fit_predict |
+--------+--------+-----+---------+---------------+------------+---------------+-------------+
| numpy  | CPU    | OK  | OK      | OK            | OK         | OK            | OK          |
| dpnp   | GPU    | OK  | OK      | OK            | OK         | OK            | OK          |
| dpnp   | CPU    | OK  | OK      | OK            | OK         | OK            | OK          |
| dpctl  | GPU    | OK  | OK      | OK            | OK         | OK            | OK          |
| dpctl  | CPU    | OK  | OK      | OK            | OK         | OK            | OK          |
| torch  | XPU    | OK  | OK      | OK            | OK         | OK            | OK          |
| torch  | CPU    | OK  | OK      | OK            | OK         | OK            | OK          |
+--------+--------+-----+---------+---------------+------------+---------------+-------------+

@wrap_output_data vs _convert_result_to_input_namespace
+---------------------------+-------------------+----------------------------------+
|                           | @wrap_output_data | _convert_result_to_input_ns      |
+---------------------------+-------------------+----------------------------------+
| Handles X=None            | No (skips)        | Yes (falls back to self._fit_X)  |
| Handles tuple returns     | No                | Yes (iterates elements)          |
| Used on kneighbors        | No (removed)      | Yes                              |
| Used on predict/score     | Yes (kept)        | Yes (for predict)                |
+---------------------------+-------------------+----------------------------------+
```

david-cortes-intel commented on doc/sources/algorithms.rst:
```diff
@@ -357,8 +361,12 @@ Classification
        - ``algorithm`` != `'brute'`
        - ``weights`` = `'callable'`
        - ``metric`` not in [`'euclidean'`, `'manhattan'`, `'minkowski'`, `'chebyshev'`, `'cosine'`]
+       - ``kd_tree`` algorithm falls back to CPU
```
It already mentions that only ``algorithm`` != `'brute'` is supported.

david-cortes-intel commented on doc/sources/array_api.rst:
```diff
@@ -149,6 +153,14 @@ that was fitted to array API inputs will work, but it will do so by transferring
 a host NumPy array, with some exceptions where |dpnp_array| classes might be returned.
 
+Similarly, :obj:`sklearn.neighbors.KNeighborsClassifier` also offers methods such as
+:meth:`~sklearn.neighbors.KNeighborsClassifier.radius_neighbors` and
```
@yuejiaointel Please merge the main branch here and add the corresponding limitations to the entries for LocalOutlierFactor in the support table, which are now present in the main branch.

yuejiaointel replied:
Thx, merged main

yuejiaointel commented on deselected_tests.yaml:
```diff
@@ -63,6 +63,12 @@ deselected_tests:
   - linear_model/tests/test_ridge.py::test_array_api_error_and_warnings_for_solver_parameter[torch]
 
+  # RidgeClassifier passes Array to xp.full() fill_value instead of Python scalar (sklearn bug with array_api_strict)
+  - linear_model/tests/test_ridge.py::test_ridge_array_api_compliance[RidgeClassifier(solver='svd')-check_array_api_attributes-array_api_strict-CPU_DEVICE-float64] >=1.8
```
rebased the changes from that pr

yuejiaointel commented on doc/sources/algorithms.rst:
yes, still making changes, tho

david-cortes-intel commented on sklearnex/neighbors/knn_unsupervised.py:
```diff
@@ -1,170 +1,251 @@
+
+
+@enable_array_api
+@control_n_jobs(decorated_methods=["fit", "kneighbors", "radius_neighbors"])
```
What about kneighbors_graph and radius_neighbors_graph?

david-cortes-intel commented on sklearnex/tests/test_common.py:
```diff
@@ -484,6 +446,17 @@ def estimator_trace(estimator, method, cache, isolated_trace):
         onedal, sklearn, or sklearnex), and callinglines is the line
         which calls the function in calledfuncs
     """
+    # Skip dynamic methods gated by available_if that are not available
```
Looks like this is conflicting with the PR that was merged adding this logic. This PR shouldn't be introducing changes.

yuejiaointel commented on doc/sources/array_api.rst:
```diff
@@ -149,6 +153,18 @@ that was fitted to array API inputs will work, but it will do so by transferring
+
+For :obj:`sklearn.neighbors.LocalOutlierFactor`, methods such as ``.fit()``,
+``.fit_predict()``, ``.score_samples()``, ``.decision_function()``, and
+``.kneighbors()`` are fully array API compatible.
```
Thx, removed

yuejiaointel commented on onedal/spmd/neighbors/neighbors.py:
```diff
@@ -36,18 +36,44 @@ def fit(self, X, y, queue=None):
+        result = self.infer(params, self._onedal_model, X_table)
+
+        responses = from_table(result.responses, like=X)
+        # Use slicing for dpctl compatibility (dpctl arrays don't have .ravel())
```
Got it, replaced with reshape

yuejiaointel commented:
/intelci: run

yuejiaointel commented:
/intelci: run

david-cortes-intel commented on onedal/utils/_array_api.py:
```diff
@@ -38,7 +38,10 @@ def _supports_buffer_protocol(obj):
     """Converted input object to array format of xp namespace provided."""
-    if hasattr(data, "__array_namespace__") or _supports_buffer_protocol(data):
+    # __dlpack__ is mandated by the array API standard and is supported by all
+    # array types we handle (numpy, dpnp, torch cpu/xpu), so a single check
```
It's not quite clear without the context of the previous commits what 'a single check' means. That part could be left out.

yuejiaointel replied:
got it, removed

david-cortes-intel commented on sklearnex/neighbors/common.py:
```diff
@@ -54,6 +261,20 @@ def _fit_validation(self, X, y=None):
 
         self.effective_metric_ = self.metric
+
+        # Only set "p" for minkowski metric, matching sklearn behavior (see PR #2945)
```
These sorts of comments could be left out.

yuejiaointel commented on sklearnex/neighbors/tests/test_neighbors.py:
```diff
@@ -83,6 +85,67 @@ def test_sklearnex_import_lof(dataframe, queue):
+@pytest.mark.parametrize("queue", get_queues())
+def test_pickle(queue):
+    if queue and queue.sycl_device.is_gpu:
+        pytest.skip("KNN classifier pickling for the GPU sycl_queue is buggy.")
```
Good catch, was old code, updated

david-cortes-intel replied:
Please test also with torch XPU arrays, to make sure that serialization of GPU models works.

yuejiaointel commented on sklearnex/neighbors/common.py:
```diff
@@ -64,15 +285,154 @@ def _fit_validation(self, X, y=None):
+        # Validate n_neighbors parameter
+        self._validate_n_neighbors(self.n_neighbors)
+
+        # Set effective metric and parameters
```
Thx, removed extra comments

david-cortes-intel commented on sklearnex/spmd/neighbors/neighbors.py:
```diff
@@ -0,0 +1,59 @@
+        """Override to call SPMD estimator predict directly."""
+        return self._onedal_estimator.predict(X, queue=queue)
+
+    def predict_proba(self, X):
```
Missing docstrings.

ethanglaser commented:
/intelci: run

yuejiaointel commented on sklearnex/neighbors/common.py:
```diff
@@ -315,10 +668,10 @@ def kneighbors_graph(self, X=None, n_neighbors=None, mode="connectivity"):
-        A_indptr = np.arange(0, n_nonzero + 1, n_neighbors)
+        A_indptr = xp.arange(0, n_nonzero + 1, n_neighbors)
 
         kneighbors_graph = sp.csr_matrix(
```
Thx, now default to csr_array and use _align_api_if_sparse to handle what user setting is

yuejiaointel commented:
/intelci: run

yuejiaointel commented:
/intelci: run

yuejiaointel commented:
> @yuejiaointel These two should output numpy, but currently output dpnp, please investigate:
> 
> ```python
> import os
> os.environ["SKLEARNEX_VERBOSE"] = "INFO"
> 
> import numpy as np
> rng = np.random.default_rng(seed=123)
> X = rng.standard_normal(size=(100,10))
> 
> import dpnp
> Xd = dpnp.array(X, device="gpu")
> 
> from sklearnex import config_context
> from sklearnex.neighbors import LocalOutlierFactor
> 
> with config_context(array_api_dispatch=False):
>     res1 = LocalOutlierFactor().fit_predict(Xd)
>     res2 = LocalOutlierFactor(novelty=True).fit(Xd).predict(Xd)
> ```
Fixed, problems is in wrap_output_data, also this is a problem for estimators
Call sequence is LocalOutlierFactor.predict(X_dpnp) -> dispatch -> transfer_to_host (convert to numpy) -> call onedal backend (numpy result back)  -> wrap_output_data -> always enter first __sycl_usm_array_interface__ check which just check if it is dpnp/dpctl, and convert back to dpnp, new change is add _array_api_offload to ensure we only do convert when dispatch is on

david-cortes-intel commented on sklearnex/neighbors/_lof.py:
```diff
@@ -145,10 +145,10 @@ def _predict(self, X=None):
-            # Array API: follow X's dtype ("everything follows X").
-            # NumPy: return int64 (sklearn convention for float64/pandas).
+            # Get namespace from result, not input, because wrap_output_data
+            # may change the result type (e.g. dpnp→numpy when dispatch is off)
```
Please avoid usage of utf8 characters like this. The system might have a different encoding where those are not supported.

ethanglaser commented on sklearnex/neighbors/knn_unsupervised.py:
```diff
@@ -1,170 +1,251 @@
+    # radius_neighbors always falls back to sklearn on CPU and returns
+    # ragged object-dtype arrays that cannot be converted to array API
+    # or GPU formats, so @wrap_output_data is intentionally omitted.
+    def radius_neighbors(
```
Does it even make sense to have this function in sklearnex?

yuejiaointel replied:
when we get torch / dpnp input, fit by onedal will return dpnp/torch input, so the internal attr like fit_x will be broken for sklearn, so we refit here to make sure we have numpy before fallback

ethanglaser commented on sklearnex/neighbors/_lof.py:
```diff
@@ -1,189 +1,227 @@
+    def fit_predict(self, X, y=None):
+        return self.fit(X)._predict()
+
+    def _kneighbors(self, X=None, n_neighbors=None, return_distance=True):
```
Should there be @wrap_output_data here?
Also - should _kneighbors be a function defined in common class or are there differences between the LOF, classification, unsupervised, regression functions? Perhaps at least the _kneighbors validation could include the validate_n_neighbors and check_is_fitted?

ethanglaser commented on sklearnex/neighbors/knn_classification.py:
```diff
@@ -1,192 +1,370 @@
+            self._onedal_estimator._onedal_model, X, params
+        )
+        xp, _ = get_namespace(X)
+        responses = from_table(result.responses, like=X)
```
In general `from_table` is used exclusively in onedal - why here? Shouldn't it be returned from onedal with from_table already applied?

ethanglaser commented on sklearnex/neighbors/knn_classification.py:
```diff
@@ -1,192 +1,370 @@
+        )
+
+    def fit(self, X, y):
+        xp, is_array_api = get_namespace(X)
```
In general why are these xp/array_api additions needed here and elsewhere throughout knn? I don't see this in other algos.

yuejiaointel commented on onedal/utils/_array_api.py:
```diff
@@ -38,7 +38,7 @@ def _supports_buffer_protocol(obj):
 def _asarray(data, xp, *args, **kwargs):
     """Converted input object to array format of xp namespace provided."""
-    if hasattr(data, "__array_namespace__") or _supports_buffer_protocol(data):
+    if hasattr(data, "__dlpack__") or _supports_buffer_protocol(data):
```
array_namespace does not work for torch xpu inputs, they don't ahve this attr, but using dlpack would work

ethanglaser commented on sklearnex/neighbors/knn_classification.py:
```diff
@@ -1,192 +1,369 @@
+        y = xp.asarray(y)
+
+        # Handle shape processing
+        shape = getattr(y, "shape", None)
```
is there any case where y would not have shape?
